fiftyone/__public__.py

```diff
@@ -7,9 +7,11 @@
 """
 import fiftyone.core.config as foc
 import fiftyone.core.service as fos
+import fiftyone.migrations as fom
 
 _database_service = fos.DatabaseService()
 config = foc.load_config()
+fom.migrate_if_necessary()
```

Comment:
This should be moved to `DatabaseService.start()` after establishing a connection. It currently blocks anything that doesn't start services (e.g. tests, documentation builds). Eventually it would also be need to moved to support #182.

Reply:
Yes, after seeing the issues with this I'm considering versioning by dataset and only forcing a migration when a dataset is loaded.

But that means the `datasets` and any other (future) meta collections would have to have a different migration strategy.

The benefit of this is that we don't force a user to migrate all of their datasets on import, which could be intensive. Thoughts?

Reply:
Yes, after seeing the issues with this I'm considering versioning by dataset and only forcing a migration when a dataset is loaded.

But that means the datasets and any other (future) meta collections would have to have a different migration strategy.

The benefit of this is that we don't force a user to migrate all of their datasets on import, which could be intensive. Thoughts?

Reply:
Yeah, doing it at dataset load time makes sense to me. The solution I suggested was just to get it to a point where I could really test this.

> But that means the datasets and any other (future) meta collections would have to have a different migration strategy.

Not sure I understand - would storing a version indicator at the dataset level not work?

Reply:
Yes, I'll use a version indicator. Migrating on dataset load should be enough, it's really only mongoengine that makes the `dataset` collection a problem. That can be fixed pretty easily, thought.

Thanks!